File: problems/0407_trapping_rain_water_ii/0407_trapping_rain_water_ii.py
class Solution:
    def trapRainWater(self, heightMap: List[List[int]]) -> int:
        m = len(heightMap)
        n = len(heightMap[0])
        visited = [[False] * n for _ in range(m)]
        pq = [] # sotred as [height, r, c]

        # Add boundary cells
        for i in range(m):
            visited[i][0] = True
            visited[i][n-1] = True
            heapq.heappush(pq, (heightMap[i][0], i, 0))
            heapq.heappush(pq, (heightMap[i][n-1], i, n-1))
        for j in range(1, n-1):
            visited[0][j] = True
            visited[m-1][j] = True
            heapq.heappush(pq, (heightMap[0][j], 0, j))
            heapq.heappush(pq, (heightMap[m-1][j], m-1, j))
        res = 0
        dirs = [(1, 0), (-1, 0), (0, 1), (0, -1)]
        while pq:
            h, r, c = heapq.heappop(pq)
            for dr, dc in dirs:
                nr, nc = r + dr, c + dc
                if nr < 0 or nr >= m or nc < 0 or nc >= n or visited[nr][nc]:
                    continue
                visited[nr][nc] = True
                nh = heightMap[nr][nc]
                if nh < h:
                    res += h - nh
                heapq.heappush(pq, (max(h, nh), nr, nc))
        return res
        

File: problems/0407_trapping_rain_water_ii/0407_trapping_rain_water_ii.java
class Solution {
    private final int[][] DIRS = {{1, 0}, {-1, 0}, {0, 1}, {0, -1}};

    public int trapRainWater(int[][] heightMap) {
        int m = heightMap.length, n = heightMap[0].length;
        boolean[][] visited = new boolean[m][n];
        PriorityQueue<int[]> pq = new PriorityQueue<>((a, b) -> a[0] - b[0]);
        // pq storead as [height, r, c]

        // add boundary cells
        for (int i = 0; i < m; i++) {
            visited[i][0] = true;
            visited[i][n-1] = true;
            pq.offer(new int[]{heightMap[i][0], i, 0});
            pq.offer(new int[]{heightMap[i][n-1], i, n-1});
        }
        for (int j = 1; j < n - 1; j++) {
            visited[0][j] = true;
            visited[m - 1][j] = true;
            pq.offer(new int[]{heightMap[0][j], 0, j});
            pq.offer(new int[]{heightMap[m-1][j], m-1, j});
        }

        int res = 0; 
        while (!pq.isEmpty()) {
            int[] cur = pq.poll();
            int h = cur[0], r = cur[1], c = cur[2];
            for (int[] d: DIRS) {
                int nr = r + d[0], nc = c + d[1];
                if (nr < 0 || nr >=m || nc < 0 || nc >= n) continue;
                if (visited[nr][nc]) continue;
                visited[nr][nc] = true;
                if (heightMap[nr][nc] < h) {
                    res += h - heightMap[nr][nc];
                }
                pq.offer(new int[]{Math.max(h, heightMap[nr][nc]), nr, nc});
            }
        }
        return res;
        
    }
}